scipy/stats/_qmc.py

```diff
@@ -1,29 +1,61 @@
 """Quasi-Monte Carlo engines and helpers."""
-import os
+from __future__ import annotations
+
 import copy
-import numbers
-from abc import ABC, abstractmethod
 import math
+import numbers
+import os
 import warnings
+from abc import ABC, abstractmethod
+from typing import (
+    Optional,
+    overload,
+    Union,
+    TYPE_CHECKING,
+    TypeVar
+)
 
 import numpy as np
 
-import scipy.stats as stats
-from scipy.stats._sobol import (
-    initialize_v, _cscramble, _fill_p_cumulative, _draw, _fast_forward,
-    _categorize, initialize_direction_numbers, _MAXDIM, _MAXBIT
-)
+if TYPE_CHECKING:
+    import numpy.typing as npt
+
 from scipy.stats._qmc_cy import (
     _cy_wrapper_centered_discrepancy,
     _cy_wrapper_wrap_around_discrepancy,
     _cy_wrapper_mixture_discrepancy,
     _cy_wrapper_l2_star_discrepancy,
     _cy_wrapper_update_discrepancy
 )
+from scipy.stats._sobol import (
+    initialize_v, _cscramble, _fill_p_cumulative, _draw, _fast_forward,
+    _categorize, initialize_direction_numbers, _MAXDIM, _MAXBIT
+)
+
+import scipy.stats as stats
+
 
 __all__ = ['scale', 'discrepancy', 'update_discrepancy',
            'QMCEngine', 'Sobol', 'Halton', 'LatinHypercube',
            'MultinomialQMC', 'MultivariateNormalQMC']
+if TYPE_CHECKING:
+    GeneratorType = TypeVar("GeneratorType", bound=Union[np.random.Generator,
+                                                         np.random.RandomState])
+
+IntNumber = Union[int, np.integer]
+DecimalNumber = Union[float, np.floating]
+
+SeedType = Optional[Union[IntNumber, np.random.Generator,
+                          np.random.RandomState]]
```

Comment:
Moving these lines to the `if TYPE_CHECKING:` block should get rid of the numpy 1.16-related `AttributeError` (https://github.com/scipy/scipy/pull/13833#issuecomment-821263620).

Reply:
Should I move all types declaration in the `if` statement, including `IntNumber` and every other types I wish to declare ?

Reply:
For statements that can be executed just fine during runtime this won't be (strictly) necessary. 
It can still be done however, but this would be more of a stylistic choice; one I don't have any strong opinions about.